design/ActorClassConfiguration.md

```diff
@@ -0,0 +1,125 @@
+
+2. Recognize and utilize certain actor class parameters to configure
+   state. While this might appeal to authors of actor classes, it has
+   the serious problem of requiring dfx (and foreign) users to
+   duplicate the special arguments as real canister arguments and
+   separate configuration arguments, leading to confusion when these
+   do not agree. To avoid this, one could strip the arguments at IDL
+   export, and bind them on constructor entry using appropriate system calls (assuming
+   these are available - probably not yet since they are expessed as Candid types).
+
+   This approach seems too much of hack.
+
+3. Extend actor class definition to introduce a second value, with a
+   derived name `Configure<class-name>` (please suggest a better
+   prefix or suffix).  This would be a curried constructor that takes
+   the explicit installation arguments and returns a specialized
+   constructor (with the usual signature) for instances of that class.
+   Naive users (and simple examples) would continue to use the default
+   class constructor; advanced users could resort to the
+   `Configure<class-name>` constructor when necessary.
+
+   For example, the declaration:
+
+```
+   class Counter(step : Nat) {
+     var state = 0;
+	 public shared inc() { state += step; };
+   }
+```
+   would introduce one type and two values:
+
+```
+   type Counter = actor { inc: () -> () };
+   func ConfigureCounter(
+     mode: { #install;
+             #upgrade : Principal;   // use actor {} instead?
+			 #reinstall : Principal; // use Counter instead?
+     },
+     compute_allocation: ?Nat;
+     memory_allocation: ?Nat)
+	 (step : Nat) :
+	 async Counter { ... }
+   func Counter(step : Nat) : async Counter; // = ConfigureCounter(#install, null, null);
+```
+
+The `Principal` argument to `#upgrade` and `#reinstall` (passed
+separately in the System API) allows the programmer to at least do (unsafe)
+programmatic upgrades of the given principals.
+
+Remark:
+
+Attempting to make this interface safer, avoiding raw principals or `actor {}`, seems to require *lower-bounds* on type parameters (which we don't have, but might be able to encode).
+
+Moreover, even if we had lower bounds, we would still not be verifying
+compatibility of _stable_ state, which we don't yet track
+in actor types and wouldn't be never be able to check for foreign actors
+anyway.
+
+Here's what a slightly safer API would look like, assuming lower-bounds. It's better to introduce three functions, since only one requires a type parameter.
+
+```
+   type Counter = actor { inc: () -> () };
+   func InstallCounter(
+     compute_allocation: ?Nat;
+     memory_allocation: ?Nat)
+	 (step : Nat) :
+   fun ReinstallCounter(
+     actor: Counter;
+     compute_allocation: ?Nat;
+     memory_allocation: ?Nat)
+	 (step : Nat) :
+	 async Counter;
+   func Upgrade<A :> Counter> { // note the lower bound (inverting `<:`)
```

Comment:
I agree!

Reply:
> I actually imagine an app like bigmap would like to upgrade its buckets, for example, without having to resort to scripting the IC somehow.

@crusso It's really interesting reading this almost 2 years in the future and I've been building rolling upgrades features and auto-scaling features for CanDB (An extension of BigMap that I've been working on for the past 3 months).


For spinning up canisters I can expose a library function that couples the `create_canister` and `install_code` methods, but there really is no substitute for the syntax that the one-liner that the class constructor exposes (when the parent/main canister is deployed containing the wasms of all child canisters).

```
// Parent.mo
import Bucket "./Bucket";
...

let bucket = Bucket.Bucket(params);
```

However, in a larger application when the parent canister has multiple actors it needs to hold, the wasm will quickly hit the 2MB upload limit, and soon after the 10MB gzipped limit (not sure if the gzipped option yet exists in Motoko?). This issue blocks the nice syntactic sugar & feature approach of using class constructor based canister instantiation.

The solution to get around this is then to create a stable variable where the actor/child wasms will be uploaded stored after the initial deployment of the parent canister, meaning that those child actor wasms need to be uploaded actively after the fact. 

This then means for spinning up a canister from the uploaded wasm, the `ic.create_canister` and `ic.install_code` methods are called one after another, with the index canister having no clue how to infer the type or candid interface of the wasm that has just been uploaded.

I realize that this comment is purely anecdotal and doesn't bring any immediate solutions, but figured that as this feature use-case is now being hit, it might revive the discussion.